src/tools/tidy/src/no_merge.rs

```diff
@@ -0,0 +1,77 @@
+//! This check makes sure that no accidental merge commits are introduced to the repository.
+//! It forbids all merge commits that are not caused by rollups/bors or subtree syncs.
+
+use std::process::Command;
+
+pub fn check(_: (), bad: &mut bool) {
+    let Ok(remote) = get_rust_lang_rust_remote() else { return; };
+    let Ok(merge_commits) = find_merge_commits(&remote) else { return; };
+
+    let mut bad_merge_commits = merge_commits.lines().filter(|commit| {
+        !(
+            // Bors is the ruler of merge commits.
+            commit.starts_with("Auto merge of")
+                || commit.starts_with("Rollup merge of")
+                // Our fellow subtrees are fine as well.
+                // This can be extended for future subtrees.
+                || commit.contains("Merge from rustc")
+                || commit.contains("merge rustc history")
+                || commit.contains("rust-analyzer")
+                || commit.contains("clippy")
+        )
+    });
+
+    if let Some(merge) = bad_merge_commits.next() {
+        tidy_error!(
+            bad,
+            "found a merge commit in the history: `{merge}`.
+To resolve the issue, see the merge commit help at the bottom of this post: https://jyn.dev/2022/09/02/git-cheats.html.
```

Comment:
this is kind of unwieldy but umm
```suggestion
To resolve the issue, see <https://jyn.dev/2022/09/02/git-cheats.html#i-made-a-merge-commit-but-now-someone-is-telling-me-merge-commits-arent-allowed-how-do-i-fix-this-please-i-already-spent-3-hours-on-this-i-dont-want-to-spend-3-more>
```

Reply:
This is beautiful

That said I will make a PR to the dev guide so we can link that instead.

Reply:
Update here: I discussed this with @jyn514 and we came to the conclusion that it would make more sense for them to write the dev guide help instead but that may take a while, so we shouldn't block it on this and just link the blog for now

Reply:
finally got around to writing the dev-guide side - I would prefer to wait for that to land if possible: https://github.com/rust-lang/rustc-dev-guide/pull/1536

https://github.com/rust-lang/rust/pull/105702 just landed adding your function to detect the remote, so you'll need to rebase over that anyway.

Reply:
I added it as a private function in `format.rs`, maybe move it in the utilities module ?